src/main/java/com/commercetools/sync/sdk2/producttypes/utils/ProductTypeTransformUtils.java

```diff
@@ -0,0 +1,118 @@
+import com.commercetools.sync.sdk2.commons.utils.ReferenceIdToKeyCache;
+import com.commercetools.sync.sdk2.services.impl.BaseTransformServiceImpl;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import javax.annotation.Nonnull;
+import org.apache.commons.lang3.StringUtils;
+
+public final class ProductTypeTransformUtils {
+
+  /**
+   * Transforms productTypes by resolving the references and map them to ProductTypeDrafts.
+   *
+   * <p>This method replaces the ids on attribute references with keys and resolves(fetch key values
+   * for the reference id's) non null and unexpanded references of the productType{@link
+   * ProductType} by using cache.
+   *
+   * <p>If the reference ids are already cached, key values are pulled from the cache, otherwise it
+   * executes the query to fetch the key value for the reference id's and store the idToKey value
+   * pair in the cache for reuse.
+   *
+   * <p>Then maps the ProductType to ProductTypeDraft by performing reference resolution considering
+   * idToKey value from the cache.
+   *
+   * @param client commercetools client.
+   * @param referenceIdToKeyCache the instance that manages cache.
+   * @param productTypes the productTypes to replace the references and attributes id's with keys.
+   * @return a new list which contains productTypeDrafts which have all their references and
+   *     attributes references resolved and already replaced with keys.
+   */
+  @Nonnull
+  public static CompletableFuture<List<ProductTypeDraft>> toProductTypeDrafts(
+      @Nonnull final ProjectApiRoot client,
+      @Nonnull final ReferenceIdToKeyCache referenceIdToKeyCache,
+      @Nonnull final List<ProductType> productTypes) {
+
+    final ProductTypeTransformServiceImpl productTypeTransformService =
+        new ProductTypeTransformServiceImpl(client, referenceIdToKeyCache);
+    return productTypeTransformService.toProductTypeDrafts(productTypes);
+  }
+
+  private static class ProductTypeTransformServiceImpl extends BaseTransformServiceImpl {
+
+    public ProductTypeTransformServiceImpl(
+        @Nonnull final ProjectApiRoot ctpClient,
+        @Nonnull final ReferenceIdToKeyCache referenceIdToKeyCache) {
+      super(ctpClient, referenceIdToKeyCache);
+    }
+
+    @Nonnull
+    public CompletableFuture<List<ProductTypeDraft>> toProductTypeDrafts(
+        @Nonnull final List<ProductType> productTypes) throws ReferenceReplacementException {
+
+      return loadNestedProductTypeReferenceKeys(productTypes)
+          .thenApply(ignore -> mapToProductTypeDrafts(productTypes, referenceIdToKeyCache));
+    }
+
+    @Nonnull
+    private CompletableFuture<Void> loadNestedProductTypeReferenceKeys(
+        @Nonnull final List<ProductType> productTypes) {
+
+      final Set<String> setOfTypeIds = new HashSet<>(collectNestedReferenceIds(productTypes));
```

Comment:
What about the direct references? `AttributeReferenceType`

Reply:
The direct references are tested also in the test `transform_ShouldReplaceProductTypeNestedAttributeReferenceIdsWithKeys`. The test is testing set of the direct references, so in the end they're testing what you wrote here. But for better clarity I also added a test that does not have set of references, but only a direct reference: https://github.com/commercetools/commercetools-sync-java/pull/981/commits/89f1f3969caed914b8aacd34c88b98a2f7ae1ea8